File: erp_flutter/lib/core/models/attendance_request.dart
import 'package:freezed_annotation/freezed_annotation.dart';

part 'attendance_request.freezed.dart';
part 'attendance_request.g.dart';

enum AttendanceRequestType {
  @JsonValue(1) correction,
  @JsonValue(2) clockIn,
  @JsonValue(3) clockOut,
  @JsonValue(4) wfh,
  @JsonValue(5) businessTrip,
}

enum AttendanceRequestStatus {
  @JsonValue(1) pending,
  @JsonValue(2) approved,
  @JsonValue(3) rejected,
  @JsonValue(4) cancelled,
}

@freezed
abstract class AttendanceRequest with _$AttendanceRequest {
  const AttendanceRequest._();

  factory AttendanceRequest({
    required String id,
    required String userId,
    required String attendanceId,
    @JsonKey(unknownEnumValue: AttendanceRequestType.correction)
    required AttendanceRequestType requestType,
    @JsonKey(unknownEnumValue: AttendanceRequestStatus.pending)
    required AttendanceRequestStatus status,
    String? reason,
    DateTime? newClockInTime,
    DateTime? newClockOutTime,
    String? approvedById,
    DateTime? approvedDate,
    String? rejectionReason,
    required DateTime createdAt,
    DateTime? updatedAt,
  }) = _AttendanceRequest;

  factory AttendanceRequest.fromJson(Map<String, dynamic> json) =>
      _$AttendanceRequestFromJson(json);

  String get statusLabel => switch (status) {
    AttendanceRequestStatus.pending   => 'Pending',
    AttendanceRequestStatus.approved  => 'Approved',
    AttendanceRequestStatus.rejected  => 'Rejected',
    AttendanceRequestStatus.cancelled => 'Cancelled',
  };
}

File: erp_api/ERP.API/Domain/Entities/AttendanceRequest.cs
namespace ERP.API.Domain.Entities;

public class AttendanceRequest
{
    public string Id { get; set; } = Guid.NewGuid().ToString();
    public string UserId { get; set; } = string.Empty;
    public User User { get; set; } = null!;
    public string AttendanceId { get; set; } = string.Empty;
    public Attendance Attendance { get; set; } = null!;
    public AttendanceRequestType RequestType { get; set; }
    public string? Reason { get; set; }
    public DateTime? NewClockInTime { get; set; }
    public DateTime? NewClockOutTime { get; set; }
    public AttendanceStatus NewStatus { get; set; }
    public AttendanceRequestStatus Status { get; set; } = AttendanceRequestStatus.Pending;
    public string? ApprovedById { get; set; }
    public User? ApprovedBy { get; set; }
    public DateTime? ApprovedDate { get; set; }
    public string? RejectionReason { get; set; }
    public DateTime CreatedAt { get; set; } = DateTime.UtcNow;
    public DateTime? UpdatedAt { get; set; }
}

public enum AttendanceRequestType
{
    Correction = 1,
    ClockIn = 2,
    ClockOut = 3,
    WFH = 4,
    BusinessTrip = 5
}

public enum AttendanceRequestStatus
{
    Pending = 1,
    Approved = 2,
    Rejected = 3,
    Cancelled = 4
}
